mio/src/event_loop.rs

```diff
@@ -0,0 +1,342 @@
+            Message::Deschedule(tok, dir) => self.deschedule(tok, dir),
+            Message::Shutdown => self.active.set(false),
+        }
+    }
+}
+
+impl LoopHandle {
+    fn send(&self, msg: Message) {
+        let mut msg_dance = Some(msg);
+
+        if CURRENT_LOOP.is_set() {
+            CURRENT_LOOP.with(|lp| {
+                if lp.id == self.id {
+                    // Need to execute all existing requests first, to ensure
+                    // that our message is processed "in order"
+                    lp.consume_queue();
+                    lp.notify(msg_dance.take().unwrap());
+                }
+            })
+        }
+
+        if let Some(msg) = msg_dance.take() {
+            self.tx
+                .send(msg)
+                .map_err(|_| ())
+                .expect("failed to send register message") // todo: handle failure
+        }
+    }
+
+    pub fn add_source(&self, source: Source) -> AddSource {
+        AddSource {
+            loop_handle: self.clone(),
+            source: Some(source),
+            id: Arc::new(AtomicUsize::new(0)),
+            scheduled: false,
+        }
+    }
+
+    fn add_source_(&self, source: Source, id: Arc<AtomicUsize>, wake: Waiter) {
+        self.send(Message::AddSource(source, id, wake));
+    }
+
+    pub fn drop_source(&self, tok: usize) {
+        self.send(Message::DropSource(tok));
+    }
+
+    pub fn schedule(&self, tok: usize, dir: Direction, wake: Waiter) {
+        self.send(Message::Schedule(tok, dir, wake));
+    }
+
+    pub fn deschedule(&self, tok: usize, dir: Direction) {
+        self.send(Message::Deschedule(tok, dir));
+    }
+
+    pub fn shutdown(&self) {
+        self.send(Message::Shutdown);
+    }
+}
+
+const ADD_SOURCE_TOKEN: usize = 0;
+
+pub struct AddSource {
+    loop_handle: LoopHandle,
+    source: Option<Source>,
+    id: Arc<AtomicUsize>,
+    scheduled: bool,
+}
+
+impl Future for AddSource {
+    type Item = usize;
+    type Error = io::Error; // TODO: integrate channel error?
+
+    fn poll(&mut self, tokens: &Tokens) -> Option<PollResult<usize, io::Error>> {
+        if self.scheduled {
+            if tokens.may_contain(&Tokens::from_usize(ADD_SOURCE_TOKEN)) {
+                let id = self.id.load(Ordering::Relaxed);
+                if id != 0 {
+                    return Some(Ok(id))
+                }
+            }
+        } else {
+            if CURRENT_LOOP.is_set() {
+                let res = CURRENT_LOOP.with(|lp| {
+                    if lp.id == self.loop_handle.id {
+                        Some(lp.add_source(self.source.take().unwrap()))
+                    } else {
+                        None
+                    }
+                });
+                if let Some(id) = res {
+                    return Some(Ok(id));
+                }
+            }
+        }
+
+        None
+    }
+
+    fn schedule(&mut self, wake: Arc<Wake>) {
+        if self.scheduled { return; }
```

Comment:
Argh, good point. That's... going to be a bit tricky...

Need to think a bit more overall about what's going on here. There may be a better way to factor this that allows us to sidestep this question.